Automattic/woocommerce-payments
# Make Order Pay work with UPE.

Fixes #1988 

#### Changes proposed in this Pull Request









A few changes to make Order Pay work with UPE:
* When creating the payment intent, we will use the order amount instead of the cart.
* Before confirming the intent, we update the payment intent with customer, level3 data and add setup_future_usage if saving the payment. This can not be done when confirming the intent. 
* Successfully confirming the intent redirects to `get_return_url( $order )`. The order is updated with the payment intent inside of `maybe_process_upe_redirect()`. We are not submitting Order Pay form.

#### Testing instructions





* With UPE enabled, create an order from WooCommerce > Orders > New Order. Or change an existing order to 'Pending Payment'
* Click on 'Customer Payment Page', You should see UPE form.
<img width="1010" alt="UPEOrderPay" src="https://user-images.githubusercontent.com/8002881/122630810-51b8da80-d08c-11eb-8121-e0aa480be4d7.png">
Note: There is a bug here, not related to UPE that always show the form, even when 'Use a new payment method' is not selected.

* Select 'Use a new payment method', fill the form but do not fill some of the inputs.
* Click on 'Pay for Order', check that you see the errors displayed.
* Fill out the form, check the box ' Save payment information to my account for future purchases.', click on 'Pay for Order', verify the order is placed.
* Check that the payment method has been saved, for example from My Account > Payment Methods.
* Check on Stripe dashboard that the payment has a customer Id and Level3 data, check the payment method is saved on the customer.
* Set the order again to Pending Payment and test with a card that requires confirmation (4000002500003155).
* Test without saving the payment method.

I have also modified a few error messages and refactored checking that the UPE form was complete.

-------------------
